zenoh/src/sample.rs

```diff
@@ -471,19 +451,67 @@ impl Sample {
         self
     }
 
+    /// Gets the key expression on which this Sample was published.
+    #[inline]
+    pub fn key_expr(&self) -> &KeyExpr<'static> {
+        &self.key_expr
+    }
+
+    /// Gets the payload of this Sample.
+    #[inline]
+    pub fn payload(&self) -> &Payload {
+        &self.payload
+    }
+
+    /// Gets the kind of this Sample.
+    #[inline]
+    pub fn kind(&self) -> SampleKind {
+        self.kind
+    }
+
+    /// Sets the kind of this Sample.
+    #[inline]
+    #[doc(hidden)]
+    #[zenoh_macros::unstable]
+    pub fn with_kind(mut self, kind: SampleKind) -> Self {
+        self.kind = kind;
+        self
+    }
+
+    /// Gets the encoding of this sample
+    #[inline]
+    pub fn encoding(&self) -> &Encoding {
+        &self.encoding
+    }
+
     /// Gets the timestamp of this Sample.
     #[inline]
-    pub fn get_timestamp(&self) -> Option<&Timestamp> {
+    pub fn timestamp(&self) -> Option<&Timestamp> {
         self.timestamp.as_ref()
     }
 
     /// Sets the timestamp of this Sample.
     #[inline]
+    #[doc(hidden)]
+    #[zenoh_macros::unstable]
     pub fn with_timestamp(mut self, timestamp: Timestamp) -> Self {
         self.timestamp = Some(timestamp);
         self
     }
 
+    /// Gets the quality of service settings this Sample was sent with.
+    #[inline]
+    pub fn qos(&self) -> &QoS {
```

Comment:
In the `put` and `delete` builders we have explicit options to set the `Priority` and `CongestionControl`. I understand that this PR only addresses the accessor pattern and it does not modify the types exposed. I wonder if we should have `priority()` and `congestion_control()` instead of a generic `QoS`?

Reply:
Why it makes a bit more sense on the `put(...)` than on the `sample`? Given the sample contains all the information associated to the data + publication?

Reply:
put(ke, payload).with_priority(...).with_congestion_control(...) corresponds to the phrase : "_publish_ this _payload_ on this _ke_ with specified priority and congestion control values". 
sample.priority() - corresponds to the phrase: "Give me priority value this sample _was published with_. Compare this for example with sample.encoding() - which corresponds to "Give me encoding (or format) of this sample". As you can see in the latter case (and in case of put) we do not need to add any extra words (except prepositions/pronouns) beyond the ones present in the function call. This is different for sample.priority(). I believe as long as we need to add extra words to described what a line of code does, we ought to add extra layers of indirection.

Reply:
I'd like to quickly remark that the API and accessors should be the same on all bindings and languages, I wouldn't add a level of indirection in C for example.

Reply:
maybe we can have something like z_sample_qos_priority(sample) ?